exact.8.in
.TH "exact" 8
.SH NAME
exact \- smtp relay authentication based on previous pop or imap authentication
.SH SYNOPSIS
.B exact [ options ]
.SH DESCRIPTION

Often you may wish to allow selecting SMTP relaying, based on a user having
previously authenticated against an IMAP or POP server.  Although this is a bit
of a kluge, it is popular because it provides simple configuration for users.

This daemon watches the logs produced by IMAP or POP files (by default @ac_cv_watch) 
for lines indicating
a user's successful authentication.  If a user successfully authenticates, then
their IP address is added to a list of addresses from which relaying should be
allowed.

This list is periodically cleaned of old entries.
.SH PERFORMANCE

Some care has been taken to make this software perform well for high volume sites.

.TP
.B \-f
run in the foreground, rather than running as a daemon.
.TP
.B \-d
produce copious debugging output.  Only has effect when the daemon is running in
the foreground.
.SH "SEE ALSO"
exim(8), imapd(8), ipopd(8)
.SH BUGS
foo
.SH AUTHOR
Doug Winter <doug@pigeonhold.com>
.SH LICENSE
This software is released under the GNU General Public License



NEWS
$Id: NEWS,v 1.9 2004/03/27 18:01:15 doug Exp $ 

EXACT 1.40

Added support for Berkeley DB files.  You may well want to use these, it should
be more efficient than text files.  May also be supported in postfix and
sendmail.

To aid distro packaging, the server order and match parameters have been moved 
out to a seperate file.  This means that the CONFIGURATION FORMAT HAS CHANGED.  

A new parameter has been added to configure, --with-server.

--------------------------------------------------------------------------------

EXACT 1.38

Ilia Kuliev reported a bug in the cyrus regular expression, which stopped it
working when hosts had PTR records.

--------------------------------------------------------------------------------

EXACT 1.37

James Harper reported a bug in the delay calculations which meant that the 
process could sometimes start thrashing and would never stop.

--------------------------------------------------------------------------------

EXACT 1.36

Added support for TEAPOP

--------------------------------------------------------------------------------

EXACT 1.35

Modified uw-imapd match expression so it supports ipop3d also.

--------------------------------------------------------------------------------

EXACT 1.34

added support for qpopper, supplied by Blair Harrison.

tidied the config file to make it more obvious how to enable a specific
MTA.

--------------------------------------------------------------------------------

EXACT 1.33

changed mode of installed files, as per suggestion from Mike Whitaker.

committed patch from Mike Whitaker, specifying DESTDIR, for easier
debianization.

Added support for courier, patches from Mike Whitaker.

Solaris bug, patch from William Armitage:

    On Solaris (or any S5r4 derived UNIX) the fork in daemonize/dofork does not
    copy the primed Alarm timer from the parent process. hence auth_clean is
    never called.

Jon R. Luini spotted a bug in how the configure script worked on freebsd, this
has been fixed.

Upgraded the build system autoconf to 2.57, to avoid spurious errors on
freebsd.

--------------------------------------------------------------------------------

EXACT 1.32

Fixed security hole in the uw-imapd regular expression, spotted by Jorey Bump.

--------------------------------------------------------------------------------

EXACT 1.31

Added support for perdition.
Clarified the tail open debug message.

--------------------------------------------------------------------------------

EXACT 1.30

Added configuration option so it will work with Cyrus.

--------------------------------------------------------------------------------

EXACT 1.20

Fixed exact's behaviour when the watch file is missing - now, if the file is
missing, it will continue checking for the files existence until it reappears.
This seems to be the best way of handling this condition.

Fixed a problem on Solaris to do with discrepancies between POSIX standard
regular expression handling.  Some standard that is.  Now uses Extended regular
expressions, which seem to be more portable.  Thanks to Jeremy Zawodny for
reporting this.

Added some new functionality to support regression testing:
    
    * specify the configuration file on the command line
    * enable broken out logfile location, so regression logs don't
      go to syslog
    * moved all constants from exact.conf into the Makefile, so the regression
      configuration can have shorter timeouts easily

--------------------------------------------------------------------------------

EXACT 1.10

Amended it to use IP addresses in the relay file - this is better if you have
clients who have no reverse DNS.

Added usage information.

--------------------------------------------------------------------------------

EXACT 1.01

ipop3d uses "Auth" instead of "Authenticated" in it's log message.  Added this 
to the default configuration file.

--------------------------------------------------------------------------------

EXACT 1.00

Initial Release


exact.c
/* $Id: exact.c,v 1.18 2004/03/27 13:05:10 doug Exp $
 * 
 * This file is part of EXACT.
 *
 * EXACT is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * EXACT is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with EXACT; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 *
 */

#include "config.h"

#include <stdio.h>
#include <stdlib.h>
#include <sys/types.h>
#include <regex.h>
#include <assert.h>
#include <unistd.h>
#include <signal.h>
#include <sys/stat.h>
#include <string.h>

#ifdef HAVE_GETOPT_H
    #include <getopt.h>
#else
    #include "getopt.h"
#endif

#include "tail.h"
#include "logger.h"
#include "match.h"
#include "conffile.h"
#include "auth.h"
#include "daemon.h"
#include "errno.h"

typedef struct {
    int foreground;
    int sleep;
    int debug;
    int preserve;
} command_line;

command_line cmd;

int onepass() {
    int i;
    int start=0;
    match_login *l;

    tail_read();
    for(i=0;i<tail_bufflen;i++) {
        if(tail_buff[i]=='\n' || tail_buff[i]=='\0') {
            tail_buff[i]='\0';
            l=match_line(tail_buff+start);
            if(l) auth_add(l->username, l->hostname);
            start=i+1;
        }
    }
    return 0;
}

void usage() {
    fprintf(stderr,"Usage: exact [-h] [-d] [-f] [-p] [-c filename]\n");
    fprintf(stderr,"       -h | --help         show this usage information\n");
    fprintf(stderr,"       -d | --debug        more than you ever want to know\n");
    fprintf(stderr,"       -f | --foreground   don't background\n");
    fprintf(stderr,"       -p | --preserve     don't remove the relay file on exit\n");
    fprintf(stderr,"       -c | --config       configuration filename\n");
    fprintf(stderr,"                           (default %s)\n", conffile_name());
    fprintf(stderr,"see the manual page exact(8) for more information\n");
}

void cmdline(int argc, char *argv[]) {
    int c;

    cmd.foreground=0;
    cmd.sleep=0;
    cmd.debug=0;
    cmd.preserve=0;
    while(1) {
        int option_index=0;
        static struct option long_options[] = {
            {"help", 0, NULL, 'h'},
            {"foreground",  0,  NULL, 'f'},
            {"sleep", 0, NULL, 's'},
            {"debug", 0, NULL, 'd'},
            {"preserver", 0, NULL, 'p'},
            {"config", 1, NULL, 'c'},
            {0,0,0,0}
        };
        c=getopt_long(argc,argv,"phsfdc:",long_options, &option_index);
        if(c==-1)
            break;
        switch(c) {
            case 'h':
                usage();
                exit(0);
            case 'f':
                cmd.foreground=1;
                break;
            case 's':
                cmd.sleep=1;
                break;
            case 'd':
                cmd.debug=1;
                break;
            case 'p':
                cmd.preserve=1;
                break;
            case 'c':
                conffile_setname(optarg);
                break;
            default:
                fprintf(stderr, "Unknown argument: %c\n",c);
                usage();
                exit(40);
                break;
        }
    }
}

void checkpid() {
    FILE *f;
    f=fopen(conffile_param("pidfile"),"r");
    if(f) {
        int opid=0;
        fscanf(f,"%d",&opid);
        fclose(f);
        if(opid) {
            int kr=kill(opid,0);
            if(kr   != -1) {
                logger(LOG_ERR, "Exact is already running, with pid %d\n", opid);
                exit(41);
            }
        }
    }
    logger(LOG_DEBUG, "exact is not already running\n");
}

void writepid() {
    FILE *f=fopen(conffile_param("pidfile"),"w");
    if(!f) {
        logger(LOG_ERR, "Cannot write to pid file %s\n", conffile_param("pidfile"));
        exit(42);
    }
    chmod(conffile_param("pidfile"),0640);
    logger(LOG_DEBUG, "Writing pid to %s\n", conffile_param("pidfile"));
    fprintf(f,"%d",(int)getpid());
    fclose(f);
}

void exit_handler(int s) {
    unlink(conffile_param("pidfile"));
    if(!cmd.preserve)
        unlink(conffile_param("authfile"));
    logger(LOG_ERR, "terminated\n");
    exit(0);
}

int main(int argc, char *argv[]) {
    int use_syslog=0;
    cmdline(argc,argv);
    logger_init(0,cmd.debug,NULL);
    conffile_read();
    conffile_check();
    checkpid();
    match_init();
    daemonize(cmd.foreground, cmd.sleep);
    auth_init();
    if(!strcmp("syslog",conffile_param("logging")))
        use_syslog=1;
    else {
        if(!strcmp("internal",conffile_param("logging")))
            use_syslog=0;
        else {
            logger(LOG_ERR, "logging parameter is neither syslog nor internal\n");
            exit(100);
        }
    }
    if(cmd.foreground) {
        logger_init(0,cmd.debug,NULL); // use stderr
    } else {
        // never debug using syslog, because you might
        // get a loop
        logger_init(use_syslog,!use_syslog && cmd.debug,conffile_param("logfile")); 
        logger(LOG_DEBUG, "Daemonized\n");
    }
    writepid();
#ifdef WITH_DB
    if(!strcmp(conffile_param("authtype"), "text"))
        auth_write_text(); // so that the file exists
#else
    auth_write_text(); // so that the file exists
#endif
    signal(1,conffile_reload);
    signal(10,auth_dump);
    signal(15,exit_handler);
    if(!tail_open()) {
        logger(LOG_ERR,"open of %s failed.  Quitting.\n", conffile_param("maillog"));
        return 2;
    }
    logger(LOG_NOTICE, "running\n");
    while(1) {
        onepass();
    }
    assert(0);
    return 0;
}

